specs/ChainSpec/NetworkSpec.md

```diff
@@ -0,0 +1,119 @@
+
+When two nodes successfully connect to each other, they broadcast the new edge to let other peers know about this connection. When a node is disconnected from other node, it should bump the nonce by 1, sign the new edge and broadcast it to let other nodes know that the connection was removed.
+
+A removed connection will be valid, if it contains valid information from the added edge it is invalidating. This will prevent peers bump nonce by more than one when deleting an edge.
+
+When node A propose an edge to B with nonce X, it will only accept it and sign it if:
+
+- X = 1 and B doesn't know about any previous edge between A and B
+- X is odd and X > Y where Y is the nonce of the edge with higher nonce between A and B known to B.
+
+## Routing Table
+
+Every node will maintain a routing table with all existing connections and relevant information to route messages. The explicit graph with all active connection is stored at all times.
+
+```rust
+struct RoutingTable {
+    /// PeerId associated for every known account id.
+    account_peers: HashMap<AccountId, AnnounceAccount>,
+    /// Active PeerId that are part of the shortest path to each PeerId.
+    pub peer_forwarding: HashMap<PeerId, HashSet<PeerId>>,
+    /// Store last update for known edges.
+    pub edges_info: HashMap<(PeerId, PeerId), Edge>,
+    /// Hash of messages that requires routing back to respective previous hop.
+    pub route_back: HashMap<CryptoHash, PeerId>,
+    /// Current view of the network. Nodes are Peers and edges are active connections.
+    raw_graph: Graph,
+}
+
+pub struct Graph {
+    adjacency: HashMap<PeerId, HashSet<PeerId>>,
+}
+```
+
+`RoutingTable` should be update accordingly the node receives updates from the network. There are four types of updates `RoutingTable` cares about:
+
+- [Adding Edge](#Add-edge)
+- [Removing Edge](#Remove-edge)
+- [Route back message](#Route-back-messages)
+- [Announce account](#Announce-account)
+
+### Routing
+
+When a node needs to send a message to another peer, it checks in the routing table if it is connected to that peer, possibly not directed but through several hops. Then it select one of the shortest path to the target peer and send a [`RoutedMessage`](Messages.md#RoutedMessage) to the first peer in the path.
+
+When it receives a `RoutedMessage`, it check if it is the target, in that case consume the body of the message, otherwise it finds a route to the target following described approach and send the message again. Is is important that before routing a message each peer check signature from original author of the message, passing a message with invalid signature can result in ban for sender. It is not required however checking the content of the message itself.
+
+Each `RoutedMessage` is equipped with a time-to-live integer. If this message is not for us, we decrement by one this field, if value is 0, we drop this message, otherwise we forward it.
```

Comment:
"If this message is not for the node processing it, it decrement the field by one before routing it; if the value is 0, the node drops the message instead of routing it."

Generally, avoid using "we" and "us".